### PR TITLE
[X86] Add missed type extension and truncation during combine

### LINKED ISSUE #66873: i32 load + zext is merged into 64-bit load for __ptr32 on amd64
## Reproduction

https://godbolt.org/z/PGK3PEf64

```cpp
__attribute((noinline))
uint64_t LoadAsLong(uint32_t* __ptr32 foo) {
	return static_cast<uint64_t>(*foo);
}
```

I expect that the `LoadAsLong` will return `42` for the test case in the godbolt listing.
But it does not.

If do not use `__ptr32`, then the `LoadAsLong` returns 42

## Investigation

Clang produces the following assembly for the `LoadAsLong` function:

```asm
        movslq  %edi, %rax
        movq    (%rax), %rax
        retq
```

the assembly loads 64 bit value but had to load 32 bits.

As far as I see llvm IR for the function is correct before isel:

```llvm
define dso_local noundef i64 @LoadAsLong(unsigned int ptr32_sptr*)(ptr addrspace(270) nocapture noundef readonly %foo) local_unnamed_addr {
entry:
  %0 = load i32, ptr addrspace(270) %foo, align 4
  %conv = zext i32 %0 to i64
  ret i64 %conv
}
```

but after isel:

```
# Machine code for function LoadAsLong(unsigned int ptr32_sptr*): IsSSA, TracksLiveness
Function Live Ins: $rdi in %0

bb.0.entry:
  liveins: $rdi
  %0:gr64 = COPY $rdi
  %1:gr32 = COPY %0.sub_32bit:gr64
  %2:gr64 = MOVSX64rr32 killed %1:gr32
  %3:gr64 = MOV64rm killed %2:gr64, 1, $noreg, 0, $noreg :: (load (s64) from %ir.foo, align 4, addrspace 270)
  $rax = COPY %3:gr64
  RET 0, $rax

# End machine code for function LoadAsLong(unsigned int ptr32_sptr*).
```

there is `load (s64) from %ir.foo, align 4, addrspace 270`

### PR DESCRIPTION
**Pull request by e-kud**
Closes #66873

### CODE REVIEW

**Comment by RKSimon** [comment]
Please can you pre-commit the tests so we can see the codegen diff?

File: llvm/test/CodeGen/X86/mixed-ptr-sizes.ll [hunk 1]
```diff
@@ -256,3 +256,44 @@ entry:
...
+; CHECK:       # %bb.0: # %entry
+; CHECK-NEXT:    movslq %ecx, %rax
+; CHECK-NEXT:    movl (%rax), %eax
+; CHECK-NEXT:    retq
+;
+; CHECK-O0-LABEL: test_load_sptr32_zext_i64:
+; CHECK-O0:       # %bb.0: # %entry
+; CHECK-O0-NEXT:    movslq %ecx, %rax
+; CHECK-O0-NEXT:    movl (%rax), %eax
+; CHECK-O0-NEXT:    movl %eax, %eax
```
Comment (Yikarus) [inline llvm/test/CodeGen/X86/mixed-ptr-sizes.ll:271]: Is this expected?

**Comment by e-kud** [comment]
> Please can you pre-commit the tests so we can see the codegen diff?

Sure. They are here https://github.com/llvm/llvm-project/pull/68219. If they are ok, I rebase after merge.

File: llvm/test/CodeGen/X86/mixed-ptr-sizes.ll (hunk 1 above)
Comment (e-kud) [reply llvm/test/CodeGen/X86/mixed-ptr-sizes.ll:271]: Yes, why not? It is `O0`. I used to think that `eax` is taken for intermediate result of load and then we need to move this results to the return register. It happens sometimes.

Comment (Yikarus) [reply llvm/test/CodeGen/X86/mixed-ptr-sizes.ll]: Got it, thanks for explanation.

**Review by RKSimon — APPROVED** [review]
LGTM - but please can you improve the commit message, as its rather vague

**Pull request merged**